File: solutions/cpp/luhn/1/luhn.cpp
#include "luhn.h"

namespace luhn {

    bool valid(std::string s) {
        int count{0};
        char ch{};
        int sum{0};
        int t{};
        for (auto it = s.rbegin(); it != s.rend(); ++it) {
            ch = *it;
            if ('0' <= ch && ch <= '9') {
                if (count % 2 == 1) {
                    t = 2 * (ch - '0');
                    sum += t;
                    if (t > 9) {
                        sum -= 9;
                    }
                } else {
                    sum += (ch - '0');
                }
                ++count;
            } else if (ch != ' ' && ch != '\n' && ch != 'r' && ch != '\t') {
                return false;
            }
        }
        if (count <= 1) {
            return false;
        }
        return sum % 10 == 0;
    }

}  // namespace luhn

File: solutions/go/luhn/1/luhn.go
package luhn

import "strings"

func Valid(id string) (b bool) {
	runes := []rune(strings.Join(strings.Fields(id), ""))
    size := len(runes)
	if size <= 1 {
        return
    }
    offset := size % 2
    sum := 0
    for i, rune := range runes {
        num := int(rune - '0')
        if num < 0 || num > 9 {
            return
        }
        if i % 2 == offset {
            doubled := num * 2
            sum += doubled
            if doubled > 9 {
                sum -= 9
            }
        } else {
            sum += num
        }
    }
    if sum % 10 == 0 {
        b = true
    }
    return
}
